src/server/debug/server/server.go

```diff
@@ -208,73 +143,418 @@ func (s *debugServer) appLogs(ctx context.Context, tw *tar.Writer) (retErr error
+		defer rp(ctx)
+		var errs error
+		for _, pod := range app.Pods {
+			if err := s.collectDescribe(ctx, dfs, app, pod); err != nil {
+				errors.JoinInto(&errs, errors.Wrapf(err, "describe pod %q", pod.Name))
+			}
+		}
+		return errs
+	}
+}
+
+func (s *debugServer) makeLogsTask(app *debug.App, rp incProgressFunc) taskFunc {
+	return func(ctx context.Context, dfs DumpFS) error {
+		defer rp(ctx)
+		if len(app.Pods) == 0 {
+			return nil
+		}
+		var errs error
+		for _, pod := range app.Pods {
+			if err := s.kubeLogs(ctx, dfs, app, pod); err != nil {
+				errors.JoinInto(&errs, errors.Wrapf(err, "collectLogs(%s)", pod.Name))
+			}
 		}
+		return errs
 	}
-	return s.forEachWorker(ctx, pipelineInfo, func(pod *v1.Pod) error {
-		return s.handleWorkerRedirect(ctx, tw, pod, collectWorker, redirect, prefix)
+}
+
+func (s *debugServer) makeLokiTask(app *debug.App, rp incProgressFunc) taskFunc {
+	return func(ctx context.Context, dfs DumpFS) (retErr error) {
+		if !s.hasLoki() {
+			return nil
+		}
+		ctx, end := log.SpanContext(ctx, "collectLokiLogs")
+		defer end(log.Errorp(&retErr))
+		defer rp(ctx)
+		var errs error
+		ctx, cf := context.WithTimeout(ctx, 10*time.Minute)
+		defer cf()
+		if app.Pipeline != nil {
+			if err := s.forEachWorkerLoki(ctx, client.NewPipeline(app.Pipeline.Project, app.Pipeline.Name), func(pod string) (retErr error) {
+				ctx, end := log.SpanContext(ctx, "forEachWorkerLoki.worker", zap.String("pod", pod))
+				defer end(log.Errorp(&retErr))
+				if err := s.collectLogsLoki(ctx, dfs, app, pod, client.PPSWorkerUserContainerName); err != nil {
+					return errors.Wrapf(err, "collect user container loki logs for pod %q", pod)
+				}
+				if err := s.collectLogsLoki(ctx, dfs, app, pod, client.PPSWorkerSidecarContainerName); err != nil {
+					return errors.Wrapf(err, "collect storage container loki logs for pod %q", pod)
+				}
+				return nil
+			}); err != nil {
+				return err
+			}
+		} else if app.Name == "pachd" {
+			if err := s.collectLogsLoki(ctx, dfs, app, s.name, "pachd"); err != nil {
+				return errors.Wrapf(err, "collect pachd container loki logs for pod %q", s.name)
+			}
+			return nil
+		} else {
+			for _, pod := range app.Pods {
+				for _, c := range pod.Containers {
+					if err := s.collectLogsLoki(ctx, dfs, app, pod.Name, c); err != nil {
+						return errors.Wrapf(err, "collect pachd container loki logs for pod %q", s.name)
+					}
+				}
+			}
+			return nil
+		}
+		return errs
+	}
+}
+
+func makeProfilesTask(server debug.Debug_DumpV2Server, apps []*debug.App) taskFunc {
+	return func(ctx context.Context, dfs DumpFS) error {
+		var errs error
+		rp := recordProgress(server, "profiles", len(apps))
+		for _, app := range apps {
+			for _, pod := range app.Pods {
+				for _, c := range pod.Containers {
+					for _, profile := range []string{"heap", "goroutine"} {
+						if err := dfs.Write(filepath.Join(appDir(app), "pods", pod.Name, c, profile), func(w io.Writer) (retErr error) {
```

Comment:
this is where i'd split into a helper function; collectProfile(pod, container, profile) or whatever.